astropy/utils/masked/core.py

```diff
@@ -0,0 +1,1086 @@
+        self._maskiter[index] = mask
+
+    def __next__(self):
+        """
+        Return the next value, or raise StopIteration.
+        """
+        out = next(self._dataiter)[...]
+        mask = next(self._maskiter)[...]
+        return self._masked.from_unmasked(out, mask, copy=False)
+
+    next = __next__
+
+
+class MaskedNDArray(Masked, np.ndarray, base_cls=np.ndarray, data_cls=np.ndarray):
+    _mask = None
+
+    info = MaskedNDArrayInfo()
+
+    def __new__(cls, *args, mask=None, **kwargs):
+        """Get data class instance from arguments and then set mask."""
+        self = super().__new__(cls, *args, **kwargs)
+        if mask is not None:
+            self.mask = mask
+        elif self._mask is None:
+            self.mask = False
+        return self
+
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(cls, **kwargs)
+        # For all subclasses we should set a default __new__ that passes on
+        # arguments other than mask to the data class, and then sets the mask.
+        if '__new__' not in cls.__dict__:
+            def __new__(newcls, *args, mask=None, **kwargs):
+                """Get data class instance from arguments and then set mask."""
+                # Need to explicitly mention classes outside of class definition.
+                self = super(cls, newcls).__new__(newcls, *args, **kwargs)
+                if mask is not None:
+                    self.mask = mask
+                elif self._mask is None:
+                    self.mask = False
+                return self
+            cls.__new__ = __new__
+
+        if 'info' not in cls.__dict__ and hasattr(cls._data_cls, 'info'):
+            new_info = type(cls.__name__+'Info',
+                            (MaskedArraySubclassInfo,
+                             cls._data_cls.info.__class__), {})
+            cls.info = new_info()
+
+    # The two pieces typically overridden.
+    @classmethod
+    def from_unmasked(cls, data, mask=None, copy=False):
+        # Note: have to override since __new__ would use ndarray.__new__
+        # which expects the shape as its first argument, not an array.
+        data = np.array(data, subok=True, copy=copy)
+        self = data.view(cls)
+        self._set_mask(mask, copy=copy)
+        return self
+
+    @property
+    def unmasked(self):
+        return super().view(self._data_cls)
+
+    @classmethod
+    def _get_masked_cls(cls, data_cls):
+        # Short-cuts
+        if data_cls is np.ndarray:
+            return MaskedNDArray
+        elif data_cls is None:  # for .view()
+            return cls
+
+        return super()._get_masked_cls(data_cls)
+
+    @property
+    def flat(self):
+        """A 1-D iterator over the Masked array.
+
+        This returns a ``MaskedIterator`` instance, which behaves the same
+        as the `~numpy.flatiter` instance returned by `~numpy.ndarray.flat`,
+        and is similar to, but not a subclass of, Python's built-in iterator
```

Comment:
is it important that it isn't? MaskedIterator does not currently inherit from anything.

Reply:
Agreed that I can just remove "but not a subclass of" (and probably should do that for `Quantity` too...)

Reply:
Alternatively, would inheriting from python's [``collections.abc.Iterable``](https://docs.python.org/3/library/collections.abc.html#collections.abc.Iterable) be beneficial? It guarantees success on ``isinstance(obj, Iterable)`` checks.

Reply:
That `isinstance` check already works (since the class defines an `__iter__()` method.

Reply:
Actually, the iterator differs from the standard python one in that assignment is allowed. I now wrote:
```
        """A 1-D iterator over the Masked array.

        This returns a ``MaskedIterator`` instance, which behaves the same
        as the `~numpy.flatiter` instance returned by `~numpy.ndarray.flat`,
        and is similar to Python's built-in iterator, except that it also
        allows assignment.
        """
```